model-optimizer/mo/utils/ir_reader/extender.py

```diff
@@ -36,7 +37,22 @@ def attr_to_list(node: Node, attribute: str):
 
     @staticmethod
     def const_shape_infer(node: Node):
+        # Check equality of old (restored from IR) and new (calculated while shape inference) input shapes
+        node['new_input_shapes'] = list()
+        for n in node.in_ports():
+            if not node.in_port(n).disconnected(): # We use such condition to handle optional inputs
+                node.new_input_shapes.append(node.in_port(n).data.get_shape())
+        assert len(node.new_input_shapes) == len(node.old_input_shapes), \
+            'Something wrong happened while {} node with type {} copy shape inference!'.format(node.name, node.type)
+        for new_input_shape, old_input_shape in zip(node.new_input_shapes, node.old_input_shapes):
+            assert np.array_equal(new_input_shape, old_input_shape), \
+                'Something wrong happened while {} node with type {} copy shape inference!'.format(node.name, node.type)
+
+        # We need to use number of input nodes instead of number of input ports to avoid errors with numbering
+        # in node.ports dictionary, where used numbers of nodes
         i = len(node.in_nodes())
-        for num in node.out_nodes():
-            node.out_node(num).shape = int64_array(node.ports[i][0])
+
+        # Set all output shapes the same as restored from IR
+        for num in node.out_ports():
+            node.out_port(num).data.set_shape(int64_array(node.ports[i][0]))
             i += 1
```

Comment:
unclear part - why we iterate through ports starting from `num_nodes` value

Reply:
Because we store values (port shapes) in `port` dictionary where keys are port numbers as we store them in IR, not in MO.